platform/macos/os_macos.mm

```diff
@@ -47,17 +47,69 @@
 #include <sys/sysctl.h>
 
 void OS_MacOS::pre_wait_observer_cb(CFRunLoopObserverRef p_observer, CFRunLoopActivity p_activiy, void *p_context) {
-	// Prevent main loop from sleeping and redraw window during modal popup display.
-	// Do not redraw when rendering is done from the separate thread, it will conflict with the OpenGL context updates.
-
-	DisplayServerMacOS *ds = (DisplayServerMacOS *)DisplayServer::get_singleton();
-	if (get_singleton()->get_main_loop() && ds && !get_singleton()->is_separate_thread_rendering_enabled() && !ds->get_is_resizing()) {
-		Main::force_redraw();
-		if (!Main::is_iterating()) { // Avoid cyclic loop.
-			Main::iteration();
+	OS_MacOS *os = static_cast<OS_MacOS *>(OS::get_singleton());
+
+	if (os->run_stage == APP_RUN_STAGE_SETUP) {
+		Error err;
+		@autoreleasepool {
+			err = Main::setup(os->execpath, os->argc, os->argv);
+		}
+
+		if (err == OK) {
+			int ret;
+			@autoreleasepool {
+				ret = Main::start();
+			}
+			if (ret == EXIT_SUCCESS) {
+				if (os->main_loop) {
+					os->run_stage = APP_RUN_STAGE_RUN;
+					@autoreleasepool {
+						os->main_loop->initialize();
+					}
+				} else {
+					os->run_stage = APP_RUN_STAGE_END_NO_CLEANUP;
+					os->should_terminate = true;
+				}
+			} else {
+				os->set_exit_code(EXIT_FAILURE);
+				os->run_stage = APP_RUN_STAGE_END_NO_CLEANUP;
+				os->should_terminate = true;
```

Comment:
When setting this, you can call `[NSApp terminate: nil]` to trigger the standard termination logic.

I notice in your updated implementation of the `applicationShouldTerminate:` delegate method, you check this and return `NSTerminateNow` which is 👍🏻

Reply:
Wont work. Terminate will never return to the main and set exit code, so it will break a lot of stuff depending on exit code.

Reply:
I see, as you would need `[NSApp run]` to return, which it doesn't 👍🏻